File: 883-car-fleet/car-fleet.java
class Solution {
    public int carFleet(int target, int[] position, int[] speed) {
        Map<Integer, Integer> speed_map = new HashMap<>();
        int num_pos = position.length;
        for (int i = 0; i < num_pos; i++) {
            speed_map.put(position[i], speed[i]);
        }

        Arrays.sort(position);

        int fleets = 0;
        double prev_target_time = -1;
        for (int i = num_pos - 1; i > -1; i--) {
            double cur_target_time = (double) (target - position[i]) / speed_map.get(position[i]);
            if (cur_target_time > prev_target_time) {
                fleets++;
                prev_target_time = cur_target_time;
            }
        }
        return fleets;
    }
}

File: 883-car-fleet/car-fleet.py
'''
    0   3   5   8   10      12
    1   3   1   4   2
'''


class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        num_pos = len(position)
        speed_map = {position[i]:speed[i] for i in range(num_pos)}  # all pos are unique
        position.sort(key=lambda x:x)

        fleets = 0
        prev_time = -1
        for pos in position[::-1]:
            cur_time = (target - pos) / speed_map[pos]
            if cur_time > prev_time:
                fleets += 1
                prev_time = cur_time

        return fleets
                
